cache/converter.go

```diff
@@ -50,22 +52,44 @@ func needsConversion(ctx context.Context, cs content.Store, desc ocispecs.Descri
 		if !images.IsLayerType(mediaType) || esgz {
 			return false, nil
 		}
+	case compression.Nydus:
+		if !images.IsLayerType(mediaType) || isNydusBlob(desc) {
+			return false, nil
+		}
 	default:
 		return false, fmt.Errorf("unknown compression type during conversion: %q", compressionType)
 	}
 	return true, nil
 }
 
+// Some compression type can't be mixed with other compression types in the same image,
+// so if `source` is this kind of layer, but the target is other compression type, we
+// should do the forced compression.
+func needsForceCompression(source ocispecs.Descriptor, target compression.Type) bool {
+	if target == compression.Nydus {
+		return !isNydusBlob(source)
+	}
+	return isNydusBlob(source)
+}
+
 // getConverter returns converter function according to the specified compression type.
 // If no conversion is needed, this returns nil without error.
-func getConverter(ctx context.Context, cs content.Store, desc ocispecs.Descriptor, comp compression.Config) (converter.ConvertFunc, error) {
+func getConverter(ctx context.Context, ref *immutableRef, desc ocispecs.Descriptor, comp compression.Config, s session.Group) (converter.ConvertFunc, error) {
+	cs := ref.cm.ContentStore
+
 	if needs, err := needsConversion(ctx, cs, desc, comp.Type); err != nil {
 		return nil, errors.Wrapf(err, "failed to determine conversion needs")
 	} else if !needs {
 		// No conversion. No need to return an error here.
 		return nil, nil
 	}
 
+	if needsForceCompression(desc, comp.Type) {
+		return func(ctx context.Context, cs content.Store, desc ocispecs.Descriptor) (*ocispecs.Descriptor, error) {
+			return doCompression(ctx, ref, comp, s)
```

Comment:
Does this trigger the whole diff calculation? This function is for converting blob in the content store so I think diff calculation should be at the outside of this function and this function should keep agnostic about refs.

Reply:
Thanks for the reply. Nydus compression type does not currently implement decompression, so`doCompression` is used here to recompute diff and write to content store for "converting" other compression types. How about moving out this logic to keep `getConverter` cleaner?